### PR TITLE
GridSample spec change proposal

### PR DESCRIPTION
**Pull request by broune**
**Description**
Proposal for updating spec on GridSample. See https://github.com/onnx/onnx/issues/4033 .

**Motivation and Context**
See https://github.com/onnx/onnx/issues/4033 .

### CODE REVIEW

**Comment by hwangdeyu** [comment]
Hi @broune, please follow the DCO required to sign-off all the commits.

File: docs/Operators.md [hunk 1]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
 #### Attributes
 
 <dl>
-<dt><tt>align_corners</tt> : int (default is 0)</dt>
-<dd>If align_corners=1, the extrema (-1 and 1) are considered as referring to the center points of the input's corner pixels. If align_corners=0, they are instead considered as referring to the corner points of the input's corner pixels, making the sampling more resolution agnostic.</dd>
+<dt><tt>align_corners</tt> : int32 (default is 0)</dt>
```
Comment (hwangdeyu) [inline docs/Operators.md:7829]: It should be `int` in Attributes.

File: docs/Operators.md [hunk 2]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+
+
 #### Inputs
 
 <dl>
-<dt><tt>X</tt> (differentiable) : T1</dt>
-<dd>4-D tensor of shape (N, C, H, W), where N is the batch size, C is the numbers of channels, H and W are the height and width of the input data.</dd>
-<dt><tt>grid</tt> (non-differentiable) : T1</dt>
-<dd>Input offset, 4-D tensor of shape (N, H_out, W_out, 2), where H_out and W_out are the height and width of grid and output, Grid specifies the sampling pixel locations normalized by the input spatial dimensions. Therefore, it should have most values in the range of [-1, 1]. If grid has values outside the range of [-1, 1], the corresponding outputs will be handled as defined by padding_mode.</dd>
+<dt><tt>input</tt> (differentiable) : T1</dt>
```
Comment (hwangdeyu) [inline docs/Operators.md:7846]: We set this input name as `X` in kernel implementation and example. So `X` should be the input1 name.

File: docs/Operators.md [hunk 3]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+
+<dt><tt>grid</tt> (non-differentiable) : T2</dt>
+<dd>4-D tensor of shape (N, H_out, W_out, 2), where H_out and W_out are the height and width of grid and output. This is the tensor that GridSample uses to determine where to do lookups in input. This parameter is allowed to be integer, but this would typically indicate a mistake as this can only access the input data at its corners. It is not allowed for N, H_out or W_out to be zero.</dd>
 </dl>
 
 #### Outputs
 
 <dl>
-<dt><tt>Y</tt> (differentiable) : T2</dt>
+<dt><tt>output</tt> (differentiable) : T1</dt>
```
Comment (hwangdeyu) [inline docs/Operators.md:7856]: `Y`, same with `X`.

**Comment by hwangdeyu** [comment]
@XiyinOSS @askhade @gramalingam Could you also take a review of this PR that detailed the `GridSample` op? thanks!

File: docs/Operators.md [hunk 4]
```diff
@@ -7611,17 +7611,213 @@ Other versions of this operator: <a href="Changelog.md#GreaterOrEqual-12">12</a>
...
+In the horizontal (left-right) dimension, the weighting of the two nearest
+points to the left equals `frac(col)` which is the fractional part of `col`
+while the weighting of the two nearest points to the right equals `1 -
+frac(col)`.
+
+For example, if `(row, col) = (1.3, 3.9)` then the interpolated value is the
+matrix product:
+
+```math
+\text{output[n,c,row',col'] } =
```
Comment (hwangdeyu) [inline docs/Operators.md:7713]: Not Sure if there are someway to display the `mathematical formula` well in markdown file.

File: docs/Operators.md (hunk 1 above)
Comment (broune) [reply docs/Operators.md:7829]: This is one of the issues pointed out in the doc. How many bits and signedness is "int"?

File: docs/Operators.md (hunk 2 above)
Comment (broune) [reply docs/Operators.md:7846]: Yep, let me fix that.

File: docs/Operators.md (hunk 4 above)
Comment (broune) [reply docs/Operators.md:7713]: On gitlab this example (or something like it, I tried a few different versions before putting this up here) displays as beautiful LaTeX rendered math. Seems like we'd want to be able to describe math in something like a ML op spec since everything is very mathematical in ML?

There are some people with thoughts on that here, there appear to be some workarounds possible on github:

  https://stackoverflow.com/questions/35498525/latex-rendering-in-readme-md-on-github

Maybe we could have an HTML mode or something where this works?

In any case, we should have an answer about how to describe math if we don't have typesetting for it. Maybe some sort of plain ascii-based math?

Comment (hwangdeyu) [reply docs/Operators.md:7713]: Hi @broune, 
After my confirmation, ONNX spec file operator.md does not support the LaTeX formula so for. But there are some plans to support it in the next half year, and some of the limitations of the operators will be fixed at the same time.
Just as you said, we might have to describe math like the way of [RNN-14](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#rnn-14) describing the math of activation functions.

File: docs/Operators.md [hunk 5]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
 <dt><tt>T2</tt> : tensor(float16), tensor(float), tensor(double)</dt>
-<dd>Constrain output types to float tensors.</dd>
+<dd>Constrain grid operand type to float tensors.</dd>
 </dl>
 
+#### Unresolved
+
+**Some concerns from the issue reported on GridSample that are not resolved here:
+
+1. We need dimensions higher than 4D for GridSample. Can the spec run
```
Comment (hwangdeyu) [inline docs/Operators.md:7873]: Thanks for the detailed and clear issue contributions. That's a very good idea for 5D support. But if we add the 5d spec description, users who are not familiar with this op would be confused if they export their model with it in 5D and failed. So it's not a good time to add it at present. And welcome to contribute the 5D kernel implementation of GridSample.

File: docs/Operators.md [hunk 6]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+2. Math is not displaying properly on Github. Should it? How should
+math be typeset properly in the ONNX spec? Presumably we want
+LaTeX-type math since ML has a lot of math in it.
+
+3. For nearest_mode, I did not follow resize(). I removed ceil and
+floor, as I don't understand when those would be used. I added
+round_to_even and put it as the default, as it seems the most balanced
+/ unbiased rounding mode of these ones.
+
+4. I set align_corners as int32, instead of int. Many other ops just
```
Comment (hwangdeyu) [inline docs/Operators.md:7886]: The `align_corners` is an attribute type `AttributeProto::INT` in the spec. It will get the Type from AttributeProto, which only includes the `int` by desgin in [here](https://github.com/onnx/onnx/blob/96516aecd4c110b0ac57eba08ac236ebf7205728/onnx/onnx.in.proto#L120). This `int` represents `int64`. Maybe it's ok to keep `int'.

File: docs/Operators.md [hunk 7]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+
+5. I disallowd integers inputs. They make sense mathematically, but
+logically it would almost always be an errro (maybe actually just
+always) to use this, as you'd hit the grid only on the corners, which
+doesn't make a lot of sense and is almost certainly something the user
+didn't intend to do and probably indicates either a bug in the user
+code or a misunderstanding of how the op works. So I think it's better
+to get an error in this case, but let me know what you think.
+
+6. I disallowed zero dimensions. Not clear what the general thinking
```
Comment (hwangdeyu) [inline docs/Operators.md:7898]: It seems the zero dimensions is rare, but it was supported in onnx https://github.com/onnx/onnx/issues/3724.

File: docs/Operators.md [hunk 8]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+is in ONNX for zero dimensions. Maybe it should be allowed? I would
+normally allow it, but it's hard to interpolate among no information,
+not clear what e.g. the border padding mode should do in this case.
```
Comment (hwangdeyu) [inline docs/Operators.md:7901]: Hi @broune , basically the spec need to consider the op implementation so we should keep the description in accordance with kernel. Do you need further help and we can complete this change proposal together?

File: docs/Operators.md [hunk 9]
```diff
@@ -7611,17 +7611,213 @@ Other versions of this operator: <a href="Changelog.md#GreaterOrEqual-12">12</a>
...
+  `denormalize(x, W, false) = ((x + 1) * W - 1) / 2` .
+
+If `align_corners` is `1` (i.e. `true`), then for the horizontal (left-right)
+dimension, `-1` refers to the center of the left-most column of pixels while `1`
+refers to the center of the right-most column of pixels. The formula for
+denormalization of a coordinate `x` is then:
+
+  `denormalize(x, W, true) = (x + 1) * (W - 1) / 2` .
+
+The normalized coordinates in `grid` are notated as `(x, y)` while the
```
Comment (liqunfu) [inline docs/Operators.md:7662]: Relation between x, y coordinate and row, col in data matrix are understandable in general. I suggest to either use x, y[, z] in op description. Avoiding using row, col shall reduce document length.

File: docs/Operators.md [hunk 10]
```diff
@@ -7611,17 +7611,213 @@ Other versions of this operator: <a href="Changelog.md#GreaterOrEqual-12">12</a>
...
+coordinates are the values in `input` of the nearest integer coordinates. This
+corresponds to the calculation
+
+`(row', col') = (round(row), round(col))` .
+
+If two integer points are equally close in a given dimension,
+corresponding to a coordinate ending in `.5`, then a rounding occurs
+according to the value of the nearest_mode attribute.
+
+For example, if `(row, col)=(1.5,3.5)`, then the coordinates used for the lookup
```
Comment (liqunfu) [inline docs/Operators.md:7696]: I do not see explicit statement on this from pytorch doc either. in pytorch:
https://github.com/pytorch/pytorch/blob/cda0d5a57b9126c6d244fdd5b02198f05c742615/aten/src/ATen/native/GridSampler.cpp#L177
std::round is used. 
However in onnxruntime, it uses std::nearbyintf which depends on the environment. 
I perfer Pytorch's deterministic approach. 
@hwangdeyu  do you know why ORT uses std::nearbyintf instead of std::round?

**Review by liqunfu — COMMENTED** [review]
Thanks for the PR. Added my comments. I am looking forward to the final PR (real patch).

File: docs/Operators.md (hunk 9 above)
Comment (gramalingam) [reply docs/Operators.md:7662]: I think the description is good in its current form. x and y refer to normalized-coordinates in the range [-1, 1], IIUC, and the denormalize function computes the row, col used to index into the input matrix. So, they are distinct.

File: docs/Operators.md [hunk 11]
```diff
@@ -7611,17 +7611,213 @@ Other versions of this operator: <a href="Changelog.md#GreaterOrEqual-12">12</a>
...
+If `align_corners` is `1` (i.e. `true`), then for the horizontal (left-right)
+dimension, `-1` refers to the center of the left-most column of pixels while `1`
+refers to the center of the right-most column of pixels. The formula for
+denormalization of a coordinate `x` is then:
+
+  `denormalize(x, W, true) = (x + 1) * (W - 1) / 2` .
+
+The normalized coordinates in `grid` are notated as `(x, y)` while the
+denormalized coordinates used for lookups are notated as `(row, col)`, matching
+PyTorch's GridSample. The concept of `x` is vertical (left-right) which
```
Comment (gramalingam) [inline docs/Operators.md:7664]: How is vertical "left-right" and horizontal  "up-down" ? Sounds confusing to me.

File: docs/Operators.md [hunk 12]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
 
+#### Unresolved
+
+**Some concerns from the issue reported on GridSample that are not resolved here:
+
+1. We need dimensions higher than 4D for GridSample. Can the spec run
+ahead of the Microsoft ONNX implementation or does it need to wait for
+implementation to be done first?
+
+2. Math is not displaying properly on Github. Should it? How should
```
Comment (gramalingam) [inline docs/Operators.md:7877]: Good point, but orthogonal to this PR. I think we should handle this aspect separately, and just focus on GridSample in this PR.

File: docs/Operators.md [hunk 13]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+math be typeset properly in the ONNX spec? Presumably we want
+LaTeX-type math since ML has a lot of math in it.
+
+3. For nearest_mode, I did not follow resize(). I removed ceil and
```
Comment (gramalingam) [inline docs/Operators.md:7881]: Haven't verified this. But I think we shouldn't change the spec unnecessarily.

File: docs/Operators.md [hunk 14]
```diff
@@ -7630,39 +7826,79 @@ This version of the operator has been available since version 16 of the default
...
+floor, as I don't understand when those would be used. I added
+round_to_even and put it as the default, as it seems the most balanced
+/ unbiased rounding mode of these ones.
+
+4. I set align_corners as int32, instead of int. Many other ops just
+say "int". Not sure what is correct here. I think the spec should say
+what the type is precisely.
+
+5. I disallowd integers inputs. They make sense mathematically, but
```
Comment (gramalingam) [inline docs/Operators.md:7890]: Makes sense

File: docs/Operators.md (hunk 13 above)
Comment (broune) [reply docs/Operators.md]: I modified this so that the proposal now no longer changes the behavior of existing GridSample, it simply documents what's there better (other than disallowing integer and complex inputs and outputs - not sure if your existing implementation supports that).

File: docs/Operators.md (hunk 12 above)
Comment (broune) [reply docs/Operators.md]: It's displaying properly now! :)

File: docs/Operators.md (hunk 10 above)
Comment (broune) [reply docs/Operators.md]: Let me know if I need to change this, but rounding towards nearest even makes sense to me as an unbiased rounding mode.

File: docs/Operators.md (hunk 9 above)
Comment (broune) [reply docs/Operators.md]: Yes, it's unfortunate that one cannot avoid using both concepts since the op inherently involves both.

File: docs/Operators.md [hunk 15]
```diff
@@ -7611,17 +7611,213 @@ Other versions of this operator: <a href="Changelog.md#GreaterOrEqual-12">12</a>
...
+allows access at coordinates beyond the size of the input.
+
+Given a 4D `input` tensor with dimensions `[N,C,H,W] ` and a 4D `grid` tensor
+with dimensions `[N,W',H',2]`, the 4D output tensor of `GridSample` has
+dimensions `[N,C,W',H']`. Conceptually, the output element at `[n,c,row',col']`
+equals `input[n,c,row,col]` where `(row, col)` is calculated from the 2-element
+vector `(x,y)` that is equal to `grid(n,row',col')`. Conceptually, that is what
+is going on. The details are more involved as described below.
+
+See also in [torch.nn.functional.grid_sample](https://pytorch.org/docs/master/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample). Currently, ONNX GridSample requires 4D tensors.
```
[the remark this answers is not in the document]
Comment (broune) [reply docs/Operators.md]: Agreed, though I left this out for this proposal/pull request for now.

File: docs/Operators.md (hunk 8 above)
Comment (broune) [reply docs/Operators.md]: I would like to know how you can handle zero dimensions. I think there's no good way to handle them, so I assume it's an error.

File: docs/Operators.md (hunk 7 above)
Comment (broune) [reply docs/Operators.md]: It's not a question of whether zero dimensions should be supported for tensors in general, that's a definite "absolutely yes". The question is whether zero dimensions makes sense for the gridsample op in particular. There I think the answer is no, though if someone really wants to, it could make sense in some of the configurations of the op. I set it to be disallowed in all cases in this proposal / pull request.

File: docs/Operators.md (hunk 5 above)
Comment (broune) [reply docs/Operators.md]: I kept it to 4D in this pull request.

File: docs/Operators.md (hunk 4 above)
Comment (broune) [reply docs/Operators.md]: This now works.

File: docs/Operators.md (hunk 3 above)
Comment (broune) [reply docs/Operators.md]: It was described as input/grid/output in this pull request. Please reopen if the contention is that it's not allowed to be called anything other than X/grid/Y.

File: docs/Operators.md (hunk 11 above)
Comment (broune) [reply docs/Operators.md]: Good catch. Fixed.

**Comment by broune** [comment]
Took a while, but this should now be done correctly in the .cc file and ready for merging / further review .
